Modules/_asynciomodule.c

```diff
@@ -193,6 +193,44 @@ is_coroutine(PyObject *coro)
     return has_it;
 }
 
+/*[clinic input]
+_asyncio.isfuture
+
+    obj: object
+
+[clinic start generated code]*/
+
+static PyObject *
+_asyncio_isfuture_impl(PyObject *module, PyObject *obj)
+/*[clinic end generated code: output=7ceb4ecaca613a86 input=3c9f9afc85c35bc5]*/
+{
+  _Py_IDENTIFIER(__class__);
+  PyObject* class = _PyObject_GetAttrId(obj, &PyId___class__);
+  if (class == NULL) {
+    return NULL;
+  }
+  _Py_IDENTIFIER(_asyncio_future_blocking);
+  int class_has_attr = _PyObject_HasAttrId(
+      class, &PyId__asyncio_future_blocking
+  );
+  Py_DECREF(class);
+  if (class_has_attr < 0) {
+    return NULL;
+  }
+  if (!class_has_attr) {
+    Py_RETURN_FALSE;
+  }
+  PyObject* obj_attr = _PyObject_GetAttrId(obj, &PyId__asyncio_future_blocking);
+  if (obj_attr == NULL) {
+    return NULL;
+  }
+  Py_DECREF(obj_attr);
+  if (obj_attr != Py_None) {
```

Comment:
While this code is OK, you shouldn't use refs to Python objects after you DECREF them, it's a bad style. Please compare first, and decref after,